### ISSUE TITLE
More field types

### ISSUE THREAD

**Issue by vpratfr** [themeum/kirki#108]
There is still room for more field types and implementations are already available in other libraries. I may submit a few PRs for that but for others who may want to work on that too, here are some sources:
- [CMB2](https://github.com/WebDevStudios/CMB2/wiki/Field-Types) and the corresponding [PHP source code for those fields](https://github.com/WebDevStudios/CMB2/blob/master/includes/CMB2_Types.php)
- [ACF](https://github.com/elliotcondon/acf) and the corresponding [PHP source code for those fields](https://github.com/elliotcondon/acf/tree/master/core/fields)

**Comment by aristath** [comment]
Which fields exactly do you think we need?

**Comment by ghost** [comment]
What about Select 2. Would be an interesting addition.
https://select2.github.io/examples.html

**Comment by padalec** [comment]
Or something like this:
http://vafpress.com/documentation/vafpress-framework/usage-sample/google-web-fonts.html

**Comment by aristath** [comment]
@cmwwebfx Select2 is possible and quite easy to do. :+1: 

@padalec The one you linked to is a composite field... meaning that it's not a single field but a section of 3 fields (if you look at the code below the screenshot on that link you'll understand what I'm talking about).
Right now you can accomplish the same result using separate, individual controls for each of the properties in there.

If you want this as a single control however, then things get a little more complicated!
You asked for this before on #65 so let me explain the main obstacle here:
A complex field like that consisting of other sub-controls is not currently easy because Kirki uses `theme_mods` to save its options in the database and one of the limitations of theme_mods and the customizer is that **you can't save arrays** in the database. You have to serialise the array on save, then unserialise it on read and so on.
What that means is that it's a bit hard to parse all the sub-controls and combine them in 1.
In order to properly do that kirki would first have to be able to use options instead of theme_mods.
This is something we're currently working on but it won't happen overnight. :smile:

**Comment by padalec** [comment]
Ohh now I got your point! :)

**Comment by ghost** [comment]
@aristath It is not a must have, but it is a nice field to be using. Suggestions only as this is what this thread was title. 
![](https://assets-cdn.github.com/images/icons/emoji/unicode/1f604.png)

**Comment by aristath** [comment]
@cmwwebfx yeah, I like the idea of select2... We'll probably add it, and since it's just JS, we could detect if any select fields have `'select2' => true,` and enable it for those fields.
No extra control, just some js.

**Comment by vpratfr** [comment]
@aristath composite fields should not be too hard once we have abstracted the core WP classes (See also #147)

**Comment by vpratfr** [comment]
To add to the list of fields, maybe something to select posts, taxonomy, terms, etc. without much setup (just pass the post type, the taxonomy name, etc. and the select is auto-populated)

**Comment by aristath** [comment]
We now have a `select2` control and we also have a method to get post lists:

``` php
Kirki_Helper::get_posts( $args );
```

`$args` accepts the same arguments as the [`get_posts`](https://codex.wordpress.org/Template_Tags/get_posts) WP function.

**Comment by aristath** [comment]
All the fields discussed above have now been implemented and we'll be adding even more pretty soon. :)
